<h1 style="Comic Sans MS; text-align: center; color: #FFC300">Boosting Algorithms Project</h1>
<h3 id="step1" style="font-family: Comic Sans MS; color: #68FF33">1. Problem statement and data collection</h3>
<p style="color: pink">In the two previous projects we saw how we could use a decision tree and then a random forest to improve the prediction of diabetes. We have reached a point where we need to improve. <b>Can boosting be the best alternative to optimize the results?</b></p>
<ul style="color: pink">
    <li>The EDA process can be found in my Github repositories <a href="https://github.com/dorianDraper/Decision-Tree-project/blob/main/src/explore.ipynb" target="_blank">Decision-Tree</a> and <a href="https://github.com/dorianDraper/Random-Forest-project/blob/main/src/rand_forests_project.ipynb" target="_blank">Random-Forest</a></li>
    <li>To optimize and improve the results I'll generate a boosting so that there is the necessary variety to enrich the prediction. I'll train it and analyze its results, trying to modify the hyperparameters that define the model with different values and analyze their impact on the final accuracy and plot the conclusions.</li>
    <li>Finally, I'll compare the predictions and choose the model with the highest prediction.</li>
</ul>

In [1]:
import pandas as pd

train_data = pd.read_csv('C:/Users/Jorge Payà/Desktop/4Geeks/DSML Bootcamp/Boosting-Algorithms/data/processed/clean_train.csv')
test_data = pd.read_csv('C:/Users/Jorge Payà/Desktop/4Geeks/DSML Bootcamp/Boosting-Algorithms/data/processed/clean_test.csv')

train_data.head()

Pregnancies  Glucose  BloodPressure  Insulin   BMI  \
0          2.0     84.0            0.0      0.0   0.0   
1          9.0    112.0           82.0      0.0  28.2   
2          1.0    139.0           46.0     83.0  28.7   
3          0.0    161.0           50.0      0.0  21.9   
4          6.0    134.0           80.0    370.0  46.2   

   DiabetesPedigreeFunction   Age  Outcome  
0                     0.304  21.0        0  
1                     1.282  50.0        1  
2                     0.654  22.0        0  
3                     0.254  65.0        0  
4                     0.238  46.0        1

In [2]:
# Split the data into training and testing sets
X_train = train_data.drop(['Outcome'], axis=1) # training data without the target variable
y_train = train_data['Outcome'] # training data target variable
X_test = test_data.drop(['Outcome'], axis=1) # testing data without the target variable
y_test = test_data['Outcome'] # testing data target variable

<h3 id="step2" style="font-family: Comic Sans MS; color: #68FF33">2. Boosting Algorithim: XGBClassifier</h3>

In [22]:
from xgboost import XGBClassifier

# Create a XGBoost classifier and add parameters
model = XGBClassifier(
    n_estimators=50,
    max_depth=3,
    learning_rate=0.1,  
    subsample=0.5,
    random_state=42,
    objective='binary:logistic'
)

# Train the model
model.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=3, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=50, n_jobs=None,
              num_parallel_tree=None, random_state=42, ...)

In [23]:
# Predict the target variable
y_pred = model.predict(X_test)

In [24]:
from sklearn.metrics import accuracy_score

# Calculate the accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))

Accuracy: 76.62%


<ol style="color: pink">
    <li>n_estimators=100, max_depth=3, learning_rate=0.1, subsample=0.5, random_state=0 >>>> The result was <b>75.97%</b></li>
    <li>n_estimators=1000, max_depth=2, learning_rate=0.05 >>>> The result was <b>72.73%</b></li> 
    <li>max_depth=3, n_estimators=1000, learning_rate=0.05 >>>> The result was <b>68.83%</b></li>
    <li>n_estimators=50, max_depth=3, learning_rate=0.1, subsample=0.5, random_state=42, objective='binary:logistic' >>>> The result was <b>76.62%</b></li>
</ol>

In [26]:
# Save the model with pickle
from pickle import dump
dump(model, open('C:/Users/Jorge Payà/Desktop/4Geeks/DSML Bootcamp/Boosting-Algorithms/models/xgboost_model_nestimators-100_maxdep-3_learnrate-0.1_randst-0_subsampl-0.5.sav', 'wb'))

In [27]:
dump(model, open('C:/Users/Jorge Payà/Desktop/4Geeks/DSML Bootcamp/Boosting-Algorithms/models/xgboost_model_obj-binary-log_nestimators-50_maxdep-3_learnrate-0.1_randst-42_subsampl-0.5.sav', 'wb'))

<div class="alert alert-block alert-success">
<b>Conclusions:</b><br>
<ul>
    <li><b>Decision Tree</b>: The result was 77% accuracy score. Ref: <a target="_blank">https://github.com/dorianDraper/Decision-Tree-project/blob/main/src/explore.ipynb</a></li>
    <li><b>Random Forests</b>: The result was 78% accuracy score. Ref: <a target="_blank">https://github.com/dorianDraper/Random-Forest-project/blob/main/src/rand_forests_project.ipynb</a></li>
    <li><b>Boosting with XGBClassifier</b>: Last, we observe a 76.62% accuracy score.</li>
</ul>
<br>Best model is <b><u>Random Forests</u></b>.
</div>